picatrix/lib/framework.py

```diff
@@ -0,0 +1,494 @@
+
+  def __init__(
+      self, fn: MagicProtocol, arguments: Optional[List[Dict[str, str]]] = None,
+      name: Optional[str] = '', suffix: Optional[str] = ''):
+    """Initialize the Picatrix Magic."""
+    self.fn = fn
+    if name:
+      self.magic_name = name
+    elif suffix:
+      self.magic_name = f'{fn.__name__}_{suffix}'
+    else:
+      self.magic_name = fn.__name__
+
+    functools.update_wrapper(self, fn)
+    self.__name__ = self.magic_name
+
+    # pylint: disable=access-member-before-definition
+    first_line = self.__doc__.split('\n')[0]
+    self.argument_parser = MagicArgumentParser(
+        prog=self.magic_name,
+        description=first_line,
+        usage=(
+            '%%%(prog)s [arguments] data\nor\n%%%%%(prog)s [arguments]\ndata'))
+
+    if arguments:
+      for argument in arguments:
+        dest_name = argument.kwargs.get('dest')
+        type_string = argument.kwargs.get('type')
+        # We need to check whether the custom argument contains an object
+        # type, if so we will need to change it into a str and store the
+        # type in the argument parser so it can be confirmed after variable
+        # expansion.
+        if dest_name and type_string:
+          if isinstance(type_string, type):
+            type_string = str(type_string).split('\'')[1]
+          if type_string not in ('str', 'int', 'bool', 'float', 'unicode'):
+            self.argument_parser.storage[dest_name] = type_string
+            argument.kwargs['type'] = str
+
+        self.argument_parser.add_argument(*argument.argc, **argument.kwargs)
+    else:
+      _add_function_arguments_to_parser(fn, self.argument_parser)
+
+    try:
+      self.argument_parser.add_argument(
+          '--bindto', '-bindto',
+          type=str,
+          default='',
+          action='store',
+          dest='bindto',
+          help='Bind the results to a variable instead of being returned.')
+    except argparse.ArgumentError:
+      pass
+
+    try:
+      self.argument_parser.add_argument(
+          'data', nargs='?',
+          help=self.argument_parser.storage.get('_data_help', ''))
+    except argparse.ArgumentError:
+      pass
+
+    self.__doc__ = self.argument_parser.format_help()
+
+  # pylint: disable=inconsistent-return-statements
+  def __call__(self, line: str, cell: Optional[str] = None) -> Union[Any, None]:
+    line_magic = cell is None
+
+    arguments = _parse_line_string(line)
+
+    try:
+      options = self.argument_parser.parse_args(arguments)
+    except KeyError as e:
+      print((
+          'Unable to parse arguments, with error: {0!s}.\n'
+          'Correct usage is: {1:s}').format(
+              e, self.argument_parser.format_help()))
+      return
+    except error.ArgParserCompleteError:
+      # When argparser ends execution but without an error this exception
+      # is raised, eg: when "-h" or help is used. In those cases we need
+      # to return without running the magic function.
+      return
```

Comment:
so... yes, we can do that, but that would also indent all the code that follows....

so instead of 
```
try:
  something
except someerror:
  something
else:
  now all the code that follows needs to be indented ....
```

instead of this one

Reply:
Got it, how about renaming exception `ArgParseCompleteError` -> `ArgParseNonZeroStatus` or something similar that indicates that this code is about handling a conditional where ArgParse exited in a way that should stop further execution.

Current name and docstring is confusing for me as:
1. Name suggests some kind of Complete Error, that sounds like a huge error
2. Docstring suggests that this Exception indicates that ArgParser exited cleanly which is (a) not an exception, (b) not true as if argparser is executed with proper arguments (i.e. not -h) then it exits cleanly without throwing this exception.

I would suggest
```
class ArgParserNonZeroStatus(Error):
  """Raised when the argument parser has exited with non-zero status (e.g. printed help message in response -h switch)"""
```

WDYT?

Reply:
done